source/common/stats/allocator_impl.cc

```diff
@@ -258,11 +258,46 @@ class TextReadoutImpl : public StatsSharedImpl<TextReadout> {
   std::string value_ ABSL_GUARDED_BY(mutex_);
 };
 
+// An implementation of Stats::CounterGroup.
+class CounterGroupImpl : public StatsSharedImpl<CounterGroup> {
+public:
+  CounterGroupImpl(StatName name, AllocatorImpl& alloc, StatName tag_extracted_name,
+                   const StatNameTagVector& stat_name_tags, size_t max_entries)
+      : StatsSharedImpl(name, alloc, tag_extracted_name, stat_name_tags), max_entries_(max_entries),
+        values_(std::make_unique<std::atomic<uint64_t>[]>(max_entries)),
+        pending_values_(std::make_unique<std::atomic<uint64_t>[]>(max_entries)) {}
+
+  void removeFromSetLockHeld() ABSL_EXCLUSIVE_LOCKS_REQUIRED(alloc_.mutex_) override {
+    const size_t count = alloc_.counter_groups_.erase(statName());
+    ASSERT(count == 1);
+  }
+
+  // Stats::TextReadout
+  void add(size_t index, uint64_t amount) override {
+    // Note that a reader may see a new value but an old pending_increment_ or
+    // used(). From a system perspective this should be eventually consistent.
+    values_[index] += amount;
+    pending_values_[index] += amount;
+    flags_ |= Flags::Used;
+  }
+  void inc(size_t index) override { add(index, 1); }
+  uint64_t latch(size_t index) override { return pending_values_[index].exchange(0); }
+  void reset(size_t index) override { values_[index] = 0; }
+  uint64_t value(size_t index) const override { return values_[index]; }
+  size_t maxEntries() const override { return max_entries_; }
+
+private:
+  const size_t max_entries_;
```

Comment:
per earlier emails, I was thinking rather than spending this 8 bytes on max_entries_, we could instead have a reference to a CounterGroupDescriptor, which would basically be a shared `vector<std::string>` or `vector<StatName>` of some sort.

This would not take any more memory, but would allow human-readable printing of stats in the admin page, and enable existing counters that are instantiated together to someday switch to this.

Reply:
Good question! For the usecase I had in mind I basically would like a map from enum to counter and since an enum is an int, this would be fine. I'd expect to do something like group.inc(ENUM_FOO) or group.inc(ENUM_BAR).

That being said, I'm happy to consider carrying around the names if that's useful. I'm not quite sure that I understand the model you're thinking of... but considering I'm new to stats, this is not shocking :) When you say a "shared vector", what would the names be shared between? (Obviously entry A and entry B in the group have different names, but that's not what you're suggesting). With the "reference to a shared descriptor" approach, is the idea that instead of passing in "size_t max_entries" to all the various find(), make() from() methods we would instead pass in this shared descriptor? Would you envision the shared descriptor would be stored in some global variable that the callers of these methods could find?

Reply:
I think `group.inc(ENUM_BAR)` is a fine model. But I was thinking about how this gets rendered in the Envoy admin console.

I'm also picturing that -- especially if you attach these quic counter-groups to hosts or clusters, there may be 10s of thousands or 100s of thousands of instances of them. But for the Quic counter-group there would only be one vector of strings holding the human-readable names.

I think the cleanest approach is that to allocate a CounterGroup you'd first need to allocate a CounterGroupDescriptor which would be constructed with a list of names, and would. hold an internal ref-count (like other stat objects).  The caller would be responsible for holding onto the CounterGroupDescriptor to pass in every time a CounterGroup was allocated. If there was a QuicContext or similar available this descriptor could be put there.

I could also imagine an approach where you pass in the vector<string> to the CounterGroup constructor, and it would do a hash lookup in the Stat::Allocator to see if that descriptor already exists so it can share it. This would be easier to use, but would involve more lookups and holding global locks longer, so I think it'd be better to use the explicit descriptor approach.

Reply:
Oh, I see! I keep thinking of metrics in the context of Chrome where there is one instance of a particular histogram, not the Envoy model where there are (potentially) N copies of a metric. Ok, that makes sense. I'll create, as you say, a CounterGroupDescriptor which will contain a vector of names Should that be the full counter name, or just the suffix? I'm guessing just the suffix, but if it's the full counter name would we use that instead of the name in the symbol table as MetricImplBase uses?

Reply:
I think the CounterGroupDescriptor should contain the leaf names only, IMO, so the descriptor can be shared across multiple instances associated with different clusters/hosts etc.

Then they can be concatenated during admin `/stats` rendering and propagation to stats sinks (e.g. Prometheus).

I haven't figured out whether StatName should be used for the leaf names, or we should just hold them in strings. We'll have to see what call-sites look like. For consistency I'd probably use StatName.

Reply:
Ok, I think I have this basically working. Please take a look. 

I also notice that I get test failures in //test/integration:stats_integration_test:
```
[ RUN      ] IpVersions/ClusterMemoryTestRunner.MemoryLargeClusterSize/IPv6
test/integration/stats_integration_test.cc:291: Failure
Expected: (m_per_cluster) <= (40000), actual: 40164 vs 40000
Stack trace:
  0x2a1a9c2: Envoy::(anonymous namespace)::ClusterMemoryTestRunner_MemoryLargeClusterSize_Test::TestBody()
  0x5c971e4: testing::internal::HandleSehExceptionsInMethodIfSupported<>()
  0x5c8716e: testing::internal::HandleExceptionsInMethodIfSupported<>()
  0x5c6f7b3: testing::Test::Run()
  0x5c703ad: testing::TestInfo::Run()
... Google Test internal frames ...
```

I don't think I understand why this is happening though. I would have expected this PR to have added a new metric type, but not added any instances of it?

Reply:
I haven't had a chancel to look deeply yet but I'd guess this is due to adding new maps for each ThreadLocalStore::ScopeImpl.

Reply:
Ah! Yes, that makes sense. Thanks! Ok, I've update the test.